[UnifiedPDF] Context menu should have selection-activated items (copy link, lookup, search-the-web)

aprotyas opened the pull request:
#### 547b742836e50083101077f7002d57a24f0e92ea
<pre>
[UnifiedPDF] Context menu should have selection-activated items (copy link, lookup, search-the-web)
<a href="https://bugs.webkit.org/show_bug.cgi?id=269330">https://bugs.webkit.org/show_bug.cgi?id=269330</a>
<a href="https://rdar.apple.com/122918090">rdar://122918090</a>

Reviewed by Tim Horton.

We track text selections as of 274032@main, making it possible to perform
actions on said selections. This patch introduces the context menu items
that are modulated by the current selection context. Namely, we add
&quot;Copy Link&quot;, &quot;Dictionary Lookup&quot;, and &quot;Search the Web&quot;. The last of
these items has a concrete action courtesy of 274583@main, and the other
two will be implemented in upcoming commits.

In doing so, we slightly refactor how we construct the context menu by
breaking it up into sections corresponding to selections, scale
adjustment, and display mode.

* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemPDFOpenWithPreview):
(WebCore::contextMenuItemPDFCopy): Deleted.
Remove the redundant definition since contextMenuItemTagCopy() returns
the same data.

* Source/WebCore/platform/LocalizedStrings.h:
WEBCORE_EXPORT some more methods since they&apos;re being called from WebKit.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):
(WebKit::UnifiedPDFPlugin::toContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::isDisplayModeContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::titleForContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::contextMenuItem const):
(WebKit::UnifiedPDFPlugin::separatorContextMenuItem const):
(WebKit::UnifiedPDFPlugin::selectionContextMenuItems const):
(WebKit::UnifiedPDFPlugin::displayModeContextMenuItems const):
(WebKit::UnifiedPDFPlugin::scaleContextMenuItems const):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):

Canonical link: <a href="https://commits.webkit.org/274611@main">https://commits.webkit.org/274611@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a4ff6d5ebd89f3049e57f31634b35865defc8a4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39597 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18576 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41953 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42132 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35497 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41903 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21466 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15905 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40171 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15660 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34268 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13582 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35222 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43409 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35950 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35553 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39341 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11866 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16015 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8860 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16063 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15672 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1607,12 +1610,12 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
-    auto addSeparator = [&] {
-        menuItems.append({ String(), 0, enumToUnderlyingType(ContextMenuItemTag::Invalid), ContextMenuItemEnablement::Disabled, ContextMenuItemHasAction::No, ContextMenuItemIsSeparator::Yes });
+    auto addSeparator = [item = separatorContextMenuItem(), &menuItems] {
+        menuItems.append(item);
```
Moving `item` is incorrect here.

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1625,27 +1628,74 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+        { contextMenuItemTagCopy(), 0, enumToUnderlyingType(ContextMenuItemTag::Copy), ContextMenuItemEnablement::Enabled, ContextMenuItemHasAction::Yes, ContextMenuItemIsSeparator::No },
+    };
+
+    if (pdfElementTypesForPluginPoint(contextMenuPoint).contains(PDFElementType::Link))
```
`contextMenuPoint` is not in the right coordinate space. Investigating.

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/063ec2d2a085a96d3f3c25231e56c43d00ff69a6

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1625,27 +1634,76 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+        return { };
+
+    Vector<PDFContextMenuItem> items {
+        { contextMenuItemTagSearchWeb(), 0, enumToUnderlyingType(ContextMenuItemTag::WebSearch), ContextMenuItemEnablement::Enabled, ContextMenuItemHasAction::Yes, ContextMenuItemIsSeparator::No },
```
These are pretty unwieldy and all have the same last 3 arguments. Consider making a helper function to create them.

aprotyas replied:
Yes good point. Let me see

aprotyas replied:
Simplified a bunch in [4a4ff6d](https://github.com/WebKit/WebKit/commit/4a4ff6d5ebd89f3049e57f31634b35865defc8a4)

aprotyas replied:
https://github.com/WebKit/WebKit/pull/24392/files#diff-ad530d257d17aa633a189e9b9afb27cdb009a5d81cbf7f42c927d8a176d8464cR1683-R1695

```cpp
PDFContextMenuItem UnifiedPDFPlugin::contextMenuItem(ContextMenuItemTag tag) const
{
    switch (tag) {
    case ContextMenuItemTag::Unknown:
    case ContextMenuItemTag::Invalid:
        return separatorContextMenuItem();
    default: {
        auto currentDisplayMode = contextMenuItemTagFromDisplayMode(m_documentLayout.displayMode());
        int state = isDisplayModeContextMenuItemTag(tag) ? currentDisplayMode == tag : 0;
        return { titleForContextMenuItemTag(tag), state, enumToUnderlyingType(tag), ContextMenuItemEnablement::Enabled, ContextMenuItemHasAction::Yes, ContextMenuItemIsSeparator::No };
    }
    }
}
```

aprotyas committed:
[UnifiedPDF] Context menu should have selection-activated items (copy link, lookup, search-the-web)
https://bugs.webkit.org/show_bug.cgi?id=269330
rdar://122918090

Reviewed by Tim Horton.

We track text selections as of 274032@main, making it possible to perform
actions on said selections. This patch introduces the context menu items
that are modulated by the current selection context. Namely, we add
"Copy Link", "Dictionary Lookup", and "Search the Web". The last of
these items has a concrete action courtesy of 274583@main, and the other
two will be implemented in upcoming commits.

In doing so, we slightly refactor how we construct the context menu by
breaking it up into sections corresponding to selections, scale
adjustment, and display mode.

* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemPDFOpenWithPreview):
(WebCore::contextMenuItemPDFCopy): Deleted.
Remove the redundant definition since contextMenuItemTagCopy() returns
the same data.

* Source/WebCore/platform/LocalizedStrings.h:
WEBCORE_EXPORT some more methods since they're being called from WebKit.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):
(WebKit::UnifiedPDFPlugin::toContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::isDisplayModeContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::titleForContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::contextMenuItem const):
(WebKit::UnifiedPDFPlugin::separatorContextMenuItem const):
(WebKit::UnifiedPDFPlugin::selectionContextMenuItems const):
(WebKit::UnifiedPDFPlugin::displayModeContextMenuItems const):
(WebKit::UnifiedPDFPlugin::scaleContextMenuItems const):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):

Canonical link: https://commits.webkit.org/274611@main
